src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.cpp

```diff
@@ -1770,617 +1796,1423 @@ HRESULT CorProfiler::RunAutoInstrumentationLoader(const ComPtr<IMetaDataEmit2>&
+    ComPtr<IUnknown> metadata_interfaces;
+    auto             hr = this->info_->GetModuleMetaData(module_id, ofRead | ofWrite, IID_IMetaDataImport2,
+                                                         metadata_interfaces.GetAddressOf());
     if (FAILED(hr))
     {
-        Logger::Warn("GenerateLoaderMethod: DefineField _isAssemblyLoaded failed");
+        Logger::Warn("GenerateLoaderType: failed to get metadata interface for ", module_id);
         return hr;
     }
 
-    // Get a TypeRef for System.Threading.Interlocked
-    mdTypeRef interlocked_type_ref;
-    hr = metadata_emit->DefineTypeRefByName(corlib_ref, WStr("System.Threading.Interlocked"), &interlocked_type_ref);
-    if (FAILED(hr))
+    const auto& metadata_import = metadata_interfaces.As<IMetaDataImport2>(IID_IMetaDataImport);
+    const auto& metadata_emit   = metadata_interfaces.As<IMetaDataEmit2>(IID_IMetaDataEmit);
+    const auto& assembly_emit   = metadata_interfaces.As<IMetaDataAssemblyEmit>(IID_IMetaDataAssemblyEmit);
+
+    MemberResolver resolver(metadata_import, metadata_emit);
+
+    mdAssemblyRef corlib_ref = mdTokenNil;
+    // We need assemblyRef only when we generate type outside of mscorlib
+    // resolver will handle if we need to use typeRef/methodRef for other mscorlib-defined types
+    // (if corlib_ref != mdTokenNil), or typeDef/methodDef can be used otherwise
+    if (module_info.assembly.name != mscorlib_assemblyName)
     {
-        Logger::Warn("GenerateLoaderMethod: DefineTypeRefByName interlocked_type_ref failed");
-        return hr;
+        mdAssemblyRef corlib_ref;
+        hr = GetCorLibAssemblyRef(assembly_emit, corAssemblyProperty, &corlib_ref);
+
+        if (FAILED(hr))
+        {
+            Logger::Warn("GenerateLoaderType: failed to define AssemblyRef to mscorlib");
+            return hr;
+        }
     }
 
-    // Create method signature for System.Threading.Interlocked::CompareExchange(int32&, int32, int32)
-    COR_SIGNATURE interlocked_compare_exchange_signature[] = {IMAGE_CEE_CS_CALLCONV_DEFAULT,
-                                                              3,
-                                                              ELEMENT_TYPE_I4,
-                                                              ELEMENT_TYPE_BYREF,
-                                                              ELEMENT_TYPE_I4,
-                                                              ELEMENT_TYPE_I4,
-                                                              ELEMENT_TYPE_I4};
-
-    mdMemberRef interlocked_compare_member_ref;
-    hr =
-        metadata_emit->DefineMemberRef(interlocked_type_ref, WStr("CompareExchange"),
-                                       interlocked_compare_exchange_signature,
-                                       sizeof(interlocked_compare_exchange_signature), &interlocked_compare_member_ref);
-    if (FAILED(hr))
+    // TypeDef/TypeRef for System.Object
+    mdToken system_object_token;
```

Comment:
For my education, what's the benefit of using scopes here? I'm by no means an expert at C++ so using scopes like this is not something I'm used to

Reply:
Very minor - it defines scope for variable, so that I will not mistakenly use them in other parts of that long function, help group code logically and make expand/collapse feature in IDE works, and hints that we may need some helper function to define as we do similar multistring-operation many time.